contrib/win32/win32compat/misc.c

```diff
@@ -1115,216 +1098,192 @@ get_machine_domain_name(wchar_t *domain, int size)
+		goto cleanup;
 	}
-	
-	for (int i = 0; i < num_user_groups; i++)
-		to_lower_case(user_groups[i]);
-
-	print_user_groups(user, user_groups, num_user_groups);
 
-	cleanup:
-		if(local_groups)
-			NetApiBufferFree(local_groups);
-
-		if(domain_groups)
-			NetApiBufferFree(domain_groups);
-
-		if(global_universal_groups)
-			NetApiBufferFree(global_universal_groups);
+	/* create custom identifier for local token source as openssh */
+	TOKEN_SOURCE token_source;
+	PCHAR token_string_const = (PCHAR) "OpenSSH";
+	LSA_STRING token_string = { (USHORT)strlen(token_string_const), (USHORT)strlen(token_string_const) + 1, token_string_const };
+	strcpy_s(token_source.SourceName, sizeof(token_source.SourceName), token_string_const);
+	if (AllocateLocallyUniqueId(&token_source.SourceIdentifier) == 0) {
+		goto cleanup;
+	}
 
-		if(dc_name_utf16)
-			NetApiBufferFree(dc_name_utf16);
-	
-		if(user_name_utf16)
-			free(user_name_utf16);
-				
-		if(user_name)
-			free(user_name);
-
-		*ngroups = num_user_groups;
-		return user_groups;
-}
+	/* initialize kerberos s4u structure with domain account */
+	USHORT s4u_size = (USHORT)(sizeof(KERB_S4U_LOGON) + wcslen(wuser) * sizeof(wchar_t));
+	s4u_logon = (PKERB_S4U_LOGON)calloc(1, s4u_size);
+	s4u_logon->MessageType = KerbS4ULogon;
+	s4u_logon->Flags = 0x8;
+	s4u_logon->ClientUpn.Length = (USHORT)(wcslen(wuser) * sizeof(wchar_t));
+	s4u_logon->ClientUpn.MaximumLength = s4u_logon->ClientUpn.Length;
+	s4u_logon->ClientUpn.Buffer = memcpy(((PBYTE)s4u_logon) +
+		sizeof(KERB_S4U_LOGON), wuser, s4u_logon->ClientUpn.Length);
+
+	/* attempt to create logon token */
+	PVOID ProfileBuffer;
+	ULONG ProfileBufferLength;
+	LUID LogonId;
+	QUOTA_LIMITS Quotas;
+	NTSTATUS SubStatus;
+	if ((LsaLogonUser(lsa_handle, &token_string, Network, lsa_authpackage, s4u_logon, s4u_size, NULL,
+		&token_source, &ProfileBuffer, &ProfileBufferLength, &LogonId, &logon_token, &Quotas, &SubStatus)) != 0) {
+		goto cleanup;
+	}
 
-/* This method will return in "group@domain" format */
-char *
-append_domain_to_groupname(char *groupname)
-{
-	if(!groupname) return NULL;
+	/* read group sids from logon token */
+	DWORD group_size = 0;
+	if (GetTokenInformation(logon_token, TokenGroups, NULL, 0, &group_size) == 0 && GetLastError() != ERROR_INSUFFICIENT_BUFFER ||
+		GetTokenInformation(logon_token, TokenGroups, (PTOKEN_GROUPS)(group_buf =
+		(PTOKEN_GROUPS)malloc(group_size)), group_size, &group_size) == 0) {
+		goto cleanup;
+	}
 
-	int len = (int) strlen(machine_domain_name) + (int) strlen(groupname) + 2;
-	char *groupname_with_domain = malloc(len);
-	if(!groupname_with_domain) {
-		error("failed to allocate memory!");
-		return NULL;
+	/* open a new connection to the lsa policy provider for group lookup */
+	LSA_OBJECT_ATTRIBUTES ObjectAttributes;
+	ZeroMemory(&ObjectAttributes, sizeof(ObjectAttributes));
+	if (LsaOpenPolicy(NULL, &ObjectAttributes, POLICY_LOOKUP_NAMES, &lsa_policy) != 0) {
```

Comment:
This requires running process as administrator.

Can you verify ssh using non-admin account (ssh non-admin-user@ip). I just want to make sure this works for both admin, non-admin users.

Reply:
I verified this works fine as a non-admin.